google-cloud-spanner/src/test/java/com/google/cloud/spanner/DatabaseClientImplTest.java

```diff
@@ -2207,6 +2211,107 @@ public void testBatchCreateSessionsPermissionDenied() {
+
+                  return null;
+                }));
+      }
+      service.shutdown();
+      assertEquals(5000, Futures.allAsList(futures).get().size());
+    }
+  }
+
+  @Test
+  public void testBatchCreateSessionsTimesOut_whenResourceExhausted() throws Exception {
+    // Simulate a minimum execution time of 2000 milliseconds for the BatchCreateSessions RPC.
+    mockSpanner.setBatchCreateSessionsExecutionTime(
+        SimulatedExecutionTime.ofMinimumAndRandomTime(2000, 0));
+    // Add a timeout for the max amount of time (60ms) that a request waits when a session is
+    // unavailable.
+    SessionPoolOptions sessionPoolOptions =
+        SessionPoolOptions.newBuilder().setAcquireSessionTimeout(Duration.ofMillis(60)).build();
+    SpannerOptions.Builder builder =
+        SpannerOptions.newBuilder()
+            .setProjectId("my-project")
+            .setChannelProvider(channelProvider)
+            .setCredentials(NoCredentials.getInstance())
+            .setSessionPoolOption(sessionPoolOptions);
+    // Set the timeout and retry settings for BatchCreateSessions to a simple
+    // single-attempt-and-timeout after 1000ms. This will ensure that session acquisition timeout of
+    // 60ms will kick for all requests before the overall request RPC timeout is breached.
+    builder
+        .getSpannerStubSettingsBuilder()
+        .batchCreateSessionsSettings()
+        .setSimpleTimeoutNoRetries(Duration.ofMillis(1000));
+
+    try (Spanner spanner = builder.build().getService()) {
+      DatabaseId databaseId = DatabaseId.of("my-project", "my-instance", "my-database");
+      DatabaseClient client = spanner.getDatabaseClient(databaseId);
+
+      ListeningExecutorService service =
+          MoreExecutors.listeningDecorator(Executors.newFixedThreadPool(1000));
+      List<ListenableFuture<Void>> futures = new ArrayList<>(5000);
+      AtomicInteger counter = new AtomicInteger();
+      for (int i = 0; i < 5000; i++) {
+        final int index = i;
+        futures.add(
+            service.submit(
+                () -> {
+                  // The following call is non-blocking and will not generate an exception.
+                  ResultSet rs = client.singleUse().executeQuery(SELECT1);
+                  // Actually trying to get any results will cause an exception.
+                  // When number of requests > MAX_SESSIONS, post setAcquireSessionTimeout
+                  // a few requests will timeout with RESOURCE_EXHAUSTED error.
+                  SpannerException e = assertThrows(SpannerException.class, rs::next);
+                  assertEquals(ErrorCode.RESOURCE_EXHAUSTED, e.getErrorCode());
+                  System.out.printf("finished test %d\n", counter.incrementAndGet());
+
+                  return null;
+                }));
+      }
+      service.shutdown();
+      assertEquals(5000, Futures.allAsList(futures).get().size());
+    }
+  }
+
```

Comment:
One thing that we are not testing here, but that is the most likely scenario to happen, is the case where:
1. `BatchCreateSessions` works as expected (that is; fast).
2. The session pool is exhausted because all the sessions have been checked out (and maybe leaked).
3. A request for another session comes in and then fails because no sessions become available.

Could we add a test for that as well. The simplest way is probably to create a session pool with just 1 or 2 sessions, and then check those sessions out by starting for example a read-only transaction and not closing it.

Reply:
* Sure there was an existing code path that was testing a similar behaviour. I have used that as a reference to add a new test `blockAndTimeoutOnPoolExhaustion_withAcquireSessionTimeout`. 
* For the existing test `blockAndTimeoutOnPoolExhaustion`, I had to explicitly set `acquireSessionTimeout` value as null since otherwise it was going into the new code path and thus failing. This kind of demonstrates how this change is a breaking change and how the customers can mitigate by setting it to `null`.